crates/rollup-boost/src/server.rs

```diff
@@ -351,12 +433,19 @@ where
         let l2_fut = self
             .l2_client
             .fork_choice_updated_v3(fork_choice_state, payload_attributes.clone());
+        let builder_healthy = matches!(self.probes.health(), Health::Healthy);
 
         // If execution mode is disabled, return the l2 client response immediately
         if self.execution_mode().is_disabled() {
             return Ok(l2_fut.await?);
         }
 
+        // If traffic to the unhealthy builder is not allowed and the builder is unhealthy,
+        if !self.allow_traffic_to_unhealthy_builder && !builder_healthy {
```

Comment:
would be happy to see a refactoring, right now there is a lot of conditions, maybe we need to bundle them somehow